### PR TITLE
Add writer scaling

### PR DESCRIPTION
**Pull request by electrum**

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/execution/TestSqlTask.java [hunk 1]
```diff
@@ -142,10 +143,10 @@ public void testSimpleQuery()
                 Optional.of(PLAN_FRAGMENT),
                 ImmutableList.of(new TaskSource(TABLE_SCAN_NODE_ID, ImmutableSet.of(SPLIT), true)),
                 createInitialEmptyOutputBuffers(PARTITIONED).withBuffer(OUT, 0).withNoMoreBufferIds());
-        assertEquals(taskInfo.getTaskStatus().getState(), TaskState.RUNNING);
+        assertThat(taskInfo.getTaskStatus().getState()).isIn(TaskState.RUNNING, TaskState.FLUSHING);
```
Comment (electrum) [inline presto-main/src/test/java/com/facebook/presto/execution/TestSqlTask.java]: This needs to be reviewed carefully and probably redone

File: presto-main/src/main/java/com/facebook/presto/execution/StageStateMachine.java [hunk 2]
```diff
@@ -148,6 +149,11 @@ public boolean transitionToRunning()
         return stageState.setIf(RUNNING, currentState -> currentState != RUNNING && !currentState.isDone());
     }
 
+    public boolean transitionToFlushing()
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/execution/StageStateMachine.java]: `transitionToRunning()` needs to be updated to not go backwards from flushing to running

File: presto-hive/src/main/java/com/facebook/presto/hive/RecordFileWriter.java [hunk 3]
```diff
@@ -114,6 +114,12 @@ public RecordFileWriter(
     }
 
     @Override
+    public long getWrittenBytes()
+    {
+        return 0;
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/RecordFileWriter.java]: Add a comment

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/ScaledWriterScheduler.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+        public synchronized boolean isDone()
+        {
+            return done;
+        }
+
+        public synchronized SettableFuture<?> create()
+        {
+            if (!done) {
+                future = SettableFuture.create();
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/ScaledWriterScheduler.java]: maybe finish the old one, but do it out side of the synchronized lock

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/ScheduleResult.java [hunk 5]
```diff
@@ -28,6 +28,7 @@
 {
     public enum BlockedReason
     {
+        SCALED_WRITER,
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/ScheduleResult.java]: WRITER_SCALING?

**Pull request closed without merging**